### DISCUSSION TITLE
VSCode extension for developing dos programs (currently uses dosbox)

### DISCUSSION THREAD

**Discussion by stuaxo** [dosemu2/dosemu2#1884] [General]
This vscode extension is interesting
https://marketplace.visualstudio.com/items?itemName=badlogicgames.dos-dev

Using dosbox it provides and environment to build dos programs, debugging over gdb works by communicating over the fake serial port provided by dosbox.

I wonder if it would be possible to use dosemu2 instead ?

I guess getting this to invoke dosbox2 instead of dosbox is straightforward, the serial port debugging looks like a little more.

**Comment by stsp** [comment]
I don't think you want to use
dosemu2 with visualstudio under
windows.

> the serial port debugging looks like a little more.

That part is really interesting.
They implemented a gdb stub:
https://github.com/badlogic/dosbox-x/tree/master/gdbstub
BUT there seems to be a bit of
cheating going on: they put it into
a dosbox-x tree and call it
"A DOSBox-x fork with modifications to support remote debugging via GDB"
but in fact its just a plain djgpp-built app:
```
--- /dev/null
+++ b/make-main.sh
@@ -0,0 +1,3 @@
+#!/bin/sh
+set -e
+~/workspaces/djgpp/bin/i586-pc-msdosdjgpp-gcc -g -Igdbstub/ gdbstub/i386-stub.c gdbstub/i386-supp.c gdbstub/main.c -o main.exe
\ No newline at end of file
```

So I believe we can just build it and
run, and it should work.
Lets ask our debugger experts like
@andrewbird and @jwt27 if they find
that gdbstub interesting. How powerful
it is compared to what libjwdpmi
presumably has, how difficult would it
be to use it outside of dosbox-x etc.

**Reply by stuaxo** [reply]
oh, I assumed this would work under Linux; it's visual studio code, not visual studio.

I use vscode sometimes in Linux, as it does some things well, but TBH I never really got on with it.
Being able to build dos apps would definitely be a fun thing to use it with.

Maybe I should have more of a poke at this if it only works on Windows since I think all the tools the vscode extension uses are cross platform ones.

**Reply by stuaxo** [reply]
Update; looks like this is X-platform, but you're right that under windows would be odd (without cygwin or mingw support at least).

**Reply by stsp** [reply]
Good to know that it can work
on linux. But I do not care (yet)
about vscode. It would however
be very interesting to try out their
gdbstub, which is obviously not
a dosbox-x mod but a standalone
app, so you can try it out under
dosemu2 right now.

Its not like I haven't thought about
a host-based debugging in dosemu2.
I did. :) For example here:
https://sourceware.org/bugzilla/show_bug.cgi?id=30007
I am trying to come up with the
glibc extension that I'll use for that
purpose. I already got patches in
binutils bugzilla, in ia16-gcc and in
a few other places. But as always,
that will take decades to complete. :)
So you have enough time to look
into this gdbstub which is already written.

**Reply by stuaxo** [reply]
Having a look at this vscode extension is definitely on my procrastination list; at some point when I'm meant to be doing something important I'll have a play with it.

Decades makes sense, it seems to take me about a year between looking at tickets on various open source projects / my own projects.

**Reply by jwt27** [reply]
> That part is really interesting. They implemented a gdb stub: https://github.com/badlogic/dosbox-x/tree/master/gdbstub BUT there seems to be a bit of cheating going on: they put it into a dosbox-x tree and call it "A DOSBox-x fork with modifications to support remote debugging via GDB" but in fact its just a plain djgpp-built app

Seems to be just a modified version of the gdb stub that comes with djgpp:
ftp://ftp.delorie.com/pub/djgpp/current/v2misc/gdbst01s.zip

Pretty basic, gets the job done, but doesn't support things like watchpoints.

To my knowledge gdb has no concept of segments or far pointers, so this is only useful for flat djgpp applications.

**Reply by stsp** [reply]
Yeah, for non-djgpp applications
you'll get no debug info anyway.
I've just found some forum where
you were saying there is no
gdbserver for dos, but libjwdpmi
does the job (or that was my
understanding):
https://comp.os.msdos.djgpp.narkive.com/6jPD9ljf/gdbserver-for-dos
But there actually is?

**Reply by jwt27** [reply]
The `gdbserver` application itself isn't available for djgpp, and I'm not sure how feasible it would be to port that.  So the only way to do remote debugging is to integrate the server code into your own program.  I wrote my own implementation to make gdb play nice with my threading system, but that `i386-stub.c` is a good starting point otherwise.

**Reply by jwt27** [reply]
Note: I posted on that thread before I knew that `i386-stub.c` existed...

**Comment by stsp** [comment]
But this is a standalone app with main(),
so it almost looks like a gdbserver, not stub...
How does gdbstub work as a standalone
app, and how it is then different from
gdbserver?

**Reply by jwt27** [reply]
I think the benefit of gdbserver is that it's able to (re)start your program, and I suppose it implements the full remote protocol including all the obscure features.

If you look at the "[standalone app](https://github.com/badlogic/dosbox-x/blob/master/gdbstub/main.c)", you'll see it's just a demo program that does nothing.  So you'd still have to link your own code with it.  And I'm not sure why this needs a separate fork since dosbox-x already has a serial null-modem driver.

**Comment by stsp** [comment]
Thanks, quite clear now. :)
